feat: wire-up RFOX change rewards address

gomesalexandre opened the pull request:
## Description

TODO:

- [x] Input
  - [x] New address bech32 validation when using manual
  - [x] Confirmed quote
  - [x] Surface current address
  - [x] Empty state when no current address - TBD after discussing with @reallybeard 
  - [x] Same address not allowed for custom addy
  - [x] Same address not allowed for portfolio addy
  - [x] Gas fee
- [x] Confirm
  - [x] Gas fee
  - [x] Implement component for current/new address (TBD in a pile with @reallybeard later in the day)
  - [x] Sign and broadcast
- [x] Status
  - [x] Handle states
  - [x] Txid link

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes https://github.com/shapeshift/web/issues/6947

## Risk
> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

Test address change e2e:

- Current reward saddress should be displayed
- New rewards address can be input either manually, or selected from the account dropdown list
- Choosing the current address as a new rewards address, either using the account dropdown or a manual address errors and doesn't allow to continue
- The manual receive address correctly catches same address as current and invalid RUNE address, reflected as errors
- Gas fees are happy both at status and confirm steps
- Tx goes through and status step is happy

### Engineering

☝🏽 

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations

☝🏽 

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)

https://jam.dev/c/b09f9e74-d9df-4ad5-8765-16b684717d59

<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"feat_wire_up_rfox_unstaking","parentHead":"8e4e647add9c4ac6b03c8fe6fb1343b041d46d32","parentPull":6957,"trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"feat_wire_up_rfox_unstaking","parentHead":"8e4e647add9c4ac6b03c8fe6fb1343b041d46d32","parentPull":6957,"trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"feat_wire_up_rfox_unstaking","parentHead":"8e4e647add9c4ac6b03c8fe6fb1343b041d46d32","parentPull":6957,"trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"feat_wire_up_rfox_unstaking","parentHead":"8e4e647add9c4ac6b03c8fe6fb1343b041d46d32","parentPull":6957,"trunk":"develop"}
```
-->


<!-- av pr metadata
This information is embedded by the av CLI when creating PRs to track the status of stacks when using Aviator. Please do not delete or edit this section of the PR.
```
{"parent":"develop","parentHead":"","trunk":"develop"}
```
-->

gomesalexandre commented on src/pages/RFOX/components/ChangeAddress/ChangeAddressConfirm.tsx:
This component is wrong - only the input part was actually implemented in the visual bits PR https://github.com/shapeshift/web/pull/6959, the confirmed part is actually a copy and paste from `<StakeConfirm />` which isn't matching the figma at all

See runtime after cleanup of ShapeShift fee and approval fees row vs. Figma 
<img width="512" alt="image" src="https://github.com/shapeshift/web/assets/17035424/0b436078-299a-4147-b6a6-2a4c88873507">
<img width="988" alt="image" src="https://github.com/shapeshift/web/assets/17035424/ef0ba17c-6c80-417b-80a3-a0ce98812588">

gomesalexandre commented on src/components/AccountDropdown/AccountDropdown.tsx:
```diff
@@ -55,6 +55,7 @@ export type AccountDropdownProps = {
   autoSelectHighestBalance?: boolean
   // Prevents accounts in the dropdown from being selected
   disabled?: boolean
+  // Hides a given set of AccountIds from the dropdown
```
rm

gomesalexandre replied:
3e57064a

gomesalexandre commented on src/pages/RFOX/components/AddressSelection.tsx:
```diff
@@ -14,31 +14,45 @@ import { useCallback, useMemo, useState } from 'react'
-}
+} & (
+  | {
+      isNewAddress: boolean
+      validateIsNewAddress: (address: string) => boolean
+    }
+  | {
+      isNewAddress?: never
+      validateIsNewAddress?: never
```
i.e both or none, but not partial

gomesalexandre commented on src/pages/RFOX/components/ChangeAddress/ChangeAddressConfirm.tsx:
```diff
@@ -7,62 +7,227 @@ import {
-    history.push(ChangeAddressRoutePaths.Status)
-  }, [history])
+  const changeAddressCard = useMemo(() => {
+    // TODO(gomes): implement current/new rune address display here
```
Remove me

gomesalexandre replied:
dda82383cc

gomesalexandre commented on src/pages/RFOX/components/ChangeAddress/ChangeAddressInput.tsx:
```diff
@@ -1,61 +1,299 @@
+    chainId: arbitrum.id,
+    query: {
+      enabled: Boolean(stakingAssetAccountAddress),
+      // TODO(gomes): unused destructurreg values isn't an omission, it's to ensure
+      // we change it to stakingInfo[3] vs. stakingInfo[2] currently after we deploy and consume the latest version of the contract
```
To be tackled as part of https://github.com/shapeshift/web/issues/6967

gomesalexandre commented on src/pages/RFOX/components/Stake/StakeInput.tsx:
```diff
@@ -532,9 +524,11 @@ export const StakeInput: React.FC<StakeInputProps & StakeRouteProps> = ({
+              }
             >
-              {errors.amountFieldInput?.message || translate('RFOX.stake')}
+              {errors.manualRuneAddress?.message || translate('RFOX.stake')}
```
This looks wrong

gomesalexandre replied:
c36cf6e211

0xApotheosis commented on src/assets/translations/en/main.json:
```diff
@@ -2453,12 +2454,21 @@
+    "newRewardAddress": "New Reward Address",
     "rewardCycleExplainer": "This will update for the next reward cycle.",
     "addressUpdateActionPending": "Change Reward Address...",
-    "addressUpdateSucces": "You have successfully updated your Thorchain Reward Address"
+    "addressUpdateSucces": "You have successfully updated your Thorchain Reward Address",
```
We can remove both of these now.

gomesalexandre replied:
16bc6a37fb

0xApotheosis commented on src/assets/translations/en/main.json:
```diff
@@ -2453,12 +2454,21 @@
-    "addressUpdateSucces": "You have successfully updated your Thorchain Reward Address"
+    "addressUpdateSucces": "You have successfully updated your Thorchain Reward Address",
+    "noPositionTitle": "No Active Position",
+    "noPositionBody": "You will be able to change your reward address here, once you have an active position.",
```
Nitpick:

```suggestion
    "noPositionBody": "You will be able to change your reward address here once you have an active position.",
```

gomesalexandre replied:
33c3c04ac821a44378d068dc1c90e39137953598

0xApotheosis commented on src/pages/RFOX/components/Stake/StakeConfirm.tsx:
```diff
@@ -439,6 +439,9 @@ export const StakeConfirm: React.FC<StakeConfirmProps & StakeRouteProps> = ({
     await handleStake() // This isn't a mistake - we invalidate as a cleanup operation before unmount to avoid current subscribers refetching with wrong args, hence making invalidation useless
     history.push(StakeRoutePaths.Status)
 
+    history.push(StakeRoutePaths.Status)
```
Duplicate lInes?

0xApotheosis commented on src/pages/RFOX/components/ChangeAddress/ChangeAddressStatus.tsx:
```diff
@@ -79,7 +107,7 @@ export const ChangeAddressStatus: React.FC<ChangeAddressRouteProps> = () => {
       )}
       <CardFooter flexDir='column' gap={2}>
-        <Button size='lg' variant='ghost'>
+        <Button as={Link} href={txLink} size='lg' variant='ghost'>
```
```suggestion
        <Button as={Link} href={txLink} size='lg' variant='ghost' isExternal>
```

gomesalexandre replied:
436f383d1dc5e58ff1f35769078fd08fce52853d

0xApotheosis reviewed (changes requested):
Looking great!

✅ Manual change address works as expected
✅ Change address from list works as expected
✅ Cannot proceed with same address, either manually or from list
✅ Current address displayed and updated on change

Some minor things to tweak before we get this in:

❌ Can proceed with manual address after clearing it

https://github.com/shapeshift/web/assets/97164662/3aabc271-7b26-489d-9572-a84e0e2b1908

❓ The "You have successfully..." text feels a little misaligned/missing padding. Should also probably have a full stop.

<img width="333" alt="Screenshot 2024-05-27 at 2 35 13 PM" src="https://github.com/shapeshift/web/assets/97164662/eef3caf9-34ce-4c82-8c50-2f08c261ab92">

gomesalexandre commented:
Nice catch @0xApotheosis - both fixed: 
> Some minor things to tweak before we get this in:
> 
> ❌ Can proceed with manual address after clearing it
> 
>  proceed-on-empty-address.mp4 

Fixed in 29db81d08b, we were too optimistic in not firing onChange when there is no (empty) rune address. Note this commit also fixes the same issue in staking:

https://jam.dev/c/a5d65acd-1487-4250-b1db-836d87728145

> 
> ❓ The "You have successfully..." text feels a little misaligned/missing padding. Should also probably have a full stop.

Fixed in b75ff3f3cf

@woodenfurniture looks like we already have two things we improved here and want to reflect in the status steps of all three stake, unstake and change address tabs: the `isExternal` on the `<Link />`, and the text alignment improvements.

While those are not functional bugs per se, that was definitely a good call on the follow-up components restaking as a follow-up cleanup PR! 🐐

0xApotheosis reviewed (approved):
Retested, looking good ser - get 'er in.
